### ISSUE TITLE
Refactor AST for type references

### ISSUE THREAD

**Issue by ColinEberhardt** [xtuc/webassemblyjs#257]
My proposal is as follows:

```
type Signature = {
   params: Array<FuncParam>,
   results: Array<Valtype>
};

type SignatureOrTypeRef = Index | Signature

type Func = {
  ...BaseNode,
  type: "Func",
  signature: SignatureOrTypeRef
  // ...
};
```

## Examples

For named type references:

```
(type $foo (func (param i32) (result i32)))
(func (type $foo))
```

becomes =>

```
    {
      "type": "Func",
      "name": {
        "type": "Identifier",
        "value": "func_0"
      },
      "signature": {
        "type": "Identifier",
        "value": "foo"
      },
      "body": []
    }
```

For type references by index, an identifier is generated

```
(type (func (param i32) (result i32)))
(func (type 0))
```

becomes =>

```
    {
      "type": "Func",
      "name": {
        "type": "Identifier",
        "value": "func_0"
      },
      "signature": {
        "type": "Identifier",
        "value": "type_0"
      },
      "body": []
    }
```

And for signatures inline:

```
(func (param i32) (result i32))
```

becomes =>

```
    {
      "type": "Func",
      "name": {
        "type": "Identifier",
        "value": "func_0"
      },
      "signature": {
        "params": [
           {
            "id": "a",
            "valtype": "i32"
          },
          {
            "id": "b",
            "valtype": "i32"
          }
        ],
        "result": [
          "i32"
        ]
       },
      "body": []
    }
```

**Comment by ColinEberhardt** [comment]
@xtuc I guess Func has a `results` array type in anticipation of multiple return values? https://github.com/WebAssembly/design/issues/1146

**Comment by xtuc** [comment]
I like your proposal. 

> results array type in anticipation of multiple return values? 

Yes.

**Comment by ColinEberhardt** [comment]
fixed in #236
